url71665d7d68_branscha_lib_scripty_tgz-pJ8-com_sdicons_scripty_ScriptyAutoReplJ8/src/com/sdicons/scripty/parser/Eval.java

```diff
@@ -0,0 +1,762 @@
+
+                    // Quick test on the number of arguments.
+                    if(lListSize != 4) throw new CommandException("The 'defun' form should have the format (defun name (<params>) <expr>).");
+                    // Name and parameters are *not* evaluated ...
+                    final Object lName = lList.get(1);
+                    final Object lParams = lList.get(2);
+                    final Object lBody = lList.get(3);
+
+                    // Do some checking.
+                    if(lName == null || !(lName instanceof String)) throw new CommandException("The first argument in the 'defun' form should evaluate to a string.");
+                    if(lParams == null || !(lParams instanceof List)) throw new CommandException("The second argument in the 'defun' form should evaluate to a list of parameters.");
+                    final List lParamList = (List) lParams;
+                    for(Object lParam:lParamList) if(!(lParam instanceof String)) throw new CommandException("The second argument in the 'defun' form, the parameter list,  should evaluate to a list of strings.");
+                    if(lBody == null) throw new CommandException("The third argument in the 'defun' form should be an expression.");
+
+                    // Create a lambda macro.
+                    final List lLambdaMacro = new ArrayList(3);
+                    lLambdaMacro.add("lambda");
+                    lLambdaMacro.add(lParams);
+                    lLambdaMacro.add(lBody);
+
+                    // Evaluate the lambda.
+                    // Bind the resulting lambda to the name GLOBALLY!
+                    final Lambda lLambda = (Lambda) eval(lLambdaMacro, aContext);
+                    aContext.getRootContext().defBinding((String) lName, lLambda);
+                    return lLambda;
+                }
+                else if("timer".equals(lCmdCandidate))
+                {
+                    if(lListSize != 2) throw new CommandException("The 'timer' form should have the format (timer expr).");
+
+                    long lStart = System.currentTimeMillis();
+                    eval(lList.get(1), aContext);
+                    long lStop = System.currentTimeMillis();
+                    return lStop - lStart;
+                }
+                else if(lCmdCandidate instanceof String && macros.hasCommand((String)lCmdCandidate))
+                {
+                    // Built-in macro call.
+                    final ICommand lMacro = macros.getCommand((String)lCmdCandidate);
+                    final List lArgs = new ArrayList(lList.size());
+                    lArgs.addAll(lList);
+
+                    try
+                    {
+                        // Macro expansion in progress!
+                        //
+                        return eval(lMacro.execute(this, aContext, lArgs.toArray()), aContext);
+                    }
+                    catch(CommandException e)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception is caught and re-thrown within the catch clause without any other operations being performed on the exception, making the catch clause unnecessary.

**Fix**: Perform additional operations on the exception that has been caught. You can wrap a standard exception into a custom exception to add additional information or special handling. When you do that, make sure to set the original exception as the cause. Otherwise, you will lose the stack trace and message of the original exception, which will make it difficult to analyze the event that caused the exception.  
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)
Similar issue at line numbers 622, 641.